### PR TITLE
feat(config): add vault provider

### PR DESCRIPTION
**Pull request by FrankYang0529**
ref: https://github.com/fermyon/spin/pull/798, https://github.com/fermyon/spin/issues/661

Content must go through a pre-merge checklist.

## Pre-Merge Content Checklist

This documentation has been checked to ensure that:

- [X] The `title, `template`, and `date` are all set
- [X] File does not use CRLF, but uses plain LF (hint: use `cat -ve <filename> | grep '^M' | wc -l` and expect 0 as a result) 
- [X] Has passed `bart check`
- [X] Has been manually tested by running in Spin/Bartholomew (hint: use `PREVIEW_MODE=1` and run `npm run styles` to update styling)
- [X] Headings are using Title Case
- [X] Code blocks have the programming language set to properly highlight syntax and the proper copy directive
- [X] Have tested with `npm run test` and resolved all errors

### CODE REVIEW

File: content/spin/configuration.md [hunk 1]
```diff
@@ -168,6 +172,62 @@ $ export SPIN_APP_API_KEY = "1234"  # Sets the `api_key` value.
...
+HTTP/1.1 200 OK
+content-length: 26
+date: Tue, 18 Oct 2022 12:34:40 GMT
+
+Got password test_password
+```
+
+## Runtime Configuration
+
+Runtime configuration contains config provider information like [vault config provier](#vault-config-provider) and wasmtime config (WIP [spin#389](https://github.com/fermyon/spin/issues/389)).
```
[the remark this answers is not in the document]
Comment (vdice) [reply content/spin/configuration.md:228]: Do we want to withhold mentioning the wasmtime config untiil that is merged/finalized?  Or maybe reword to mention that other forms of runtime config (eg wasmtime) are on the roadmap/being planned?

**Review by vdice — COMMENTED** [review]
Thank you for this docs update!

A few grammatical suggestions while this is in draft (pending merge of https://github.com/fermyon/spin/pull/798 as well)...

@flynnduism @mikkelhegn it looks like we're tracking changes to `static/data.json` -- does this look right to you?

**Comment by karthik2804** [comment]
@vdice We are tracking `static/data.json` because we need it to be updated for search indexing.

**Comment by mikkelhegn** [comment]
@FrankYang0529 can you run `npm run build-index` to update the search index and then commit the updated data.json file? We're fixing the PR template to include this step.

File: content/spin/configuration.md [hunk 2]
```diff
@@ -156,8 +156,12 @@ api_key = "{{ api_key }}"
 ### Custom Config Providers
 
 [Custom config variables](#custom-config-variables) values may be set at runtime by
-config "providers". Currently there is only one provider: the environment
-variable provider, which gets config values from the `spin` process's
+config "providers". Currently there are providers: the environment
```
[the remark this answers is not in the document]
Comment (FrankYang0529) [reply content/spin/configuration.md]: Fixed it. Thank you.

File: content/spin/configuration.md [hunk 3]
```diff
@@ -168,6 +172,62 @@ $ export SPIN_APP_API_KEY = "1234"  # Sets the `api_key` value.
 $ spin up
 ```
 
+#### Vault Config Provider
+
+The vault config provider gets secret values from the [HashiCorp Vault](https://www.vaultproject.io/).
```
[the remark this answers is not in the document]
Comment (FrankYang0529) [reply content/spin/configuration.md]: Fixed it. Thank you.

File: content/spin/configuration.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (FrankYang0529) [reply content/spin/configuration.md]: I fixed the typo and removed the wasmtime config here. I will add it after it's implemented. Thanks.

**Comment by FrankYang0529** [comment]
> @FrankYang0529 can you run `npm run build-index` to update the search index and then commit the updated data.json file? We're fixing the PR template to include this step.

Hi @mikkelhegn, I run the command and committed it.

I have a question: how to install `bart` command? There is a check item in the list. Thank you!

**Comment by mikkelhegn** [comment]
Instructions are here: https://bartholomew.fermyon.dev/quickstart

**Comment by FrankYang0529** [comment]
> Instructions are here: https://bartholomew.fermyon.dev/quickstart

Thanks! I also add the reference to pull_request_template.

**Review by mikkelhegn — APPROVED** [review]
I haven't tested the functionality, but I trust the instructions are ok :-)

**Review by vdice — COMMENTED** [review]
Last thought: We don't yet have a good story around delineating between docs around unreleased Spin changes and docs that are valid for the most recent release (eg v0.6.0 currently).  Until then, should we add a note somewhere that the vault config provider functionality is available in the `canary` release now and anticipated to be included in the next v0.7.0 release?

(Filed https://github.com/fermyon/developer/issues/195 around docs versioning strategy)

**Comment by FrankYang0529** [comment]
> Last thought: We don't yet have a good story around delineating between docs around unreleased Spin changes and docs that are valid for the most recent release (eg v0.6.0 currently). Until then, should we add a note somewhere that the vault config provider functionality is available in the `canary` release now and anticipated to be included in the next v0.7.0 release?

Added a note for it. Thanks.

**Review by vdice — APPROVED** [review]
Thank you @FrankYang0529!
